testsuite/tests/basic/patmatch_for_multiple.ml

```diff
@@ -3,57 +3,197 @@
+
+(* Regression test for #3780 *)
+let _ = fun a b ->
+  match a, b with
+  | ((true, _) as _g)
+  | ((false, _) as _g) -> ()
+[%%expect{|
+(function a/98 b/99 0)
+- : bool -> 'a -> unit = <fun>
+|}];;
+
+(* More complete tests.
+
+   The test cases below compare the compiler output on alias patterns
+   that are outside an or-pattern (handled during half-simplification,
+   then flattened) or inside an or-pattern (handled during simplification).
+
+   We used to have a Cannot_flatten exception that would result in fairly
+   different code generated in both cases, but now the compilation strategy
+   is fairly similar.
+*)
+let _ = fun a b -> match a, b with
+| (true, _) as p -> p
+| (false, _) as p -> p
+(* outside, trivial *)
+[%%expect {|
+(function a/102 b/103 (let (p/104 =a (makeblock 0 a/102 b/103)) p/104))
+- : bool -> 'a -> bool * 'a = <fun>
+|}]
+
+let _ = fun a b -> match a, b with
+| ((true, _) as p)
+| ((false, _) as p) -> p
+(* inside, trivial *)
+[%%expect{|
+(function a/106 b/107 (let (p/108 =a (makeblock 0 a/106 b/107)) p/108))
+- : bool -> 'a -> bool * 'a = <fun>
+|}];;
+
+let _ = fun a b -> match a, b with
+| (true as x, _) as p -> x, p
+| (false as x, _) as p -> x, p
+(* outside, simple *)
+[%%expect {|
+(function a/112 b/113
+  (let (x/114 =a a/112 p/115 =a (makeblock 0 a/112 b/113))
+    (makeblock 0 x/114 p/115)))
+- : bool -> 'a -> bool * (bool * 'a) = <fun>
+|}]
+
+let _ = fun a b -> match a, b with
+| ((true as x, _) as p)
+| ((false as x, _) as p) -> x, p
+(* inside, simple *)
+[%%expect {|
+(function a/118 b/119
+  (let (x/120 =a a/118 p/121 =a (makeblock 0 a/118 b/119))
+    (makeblock 0 x/120 p/121)))
+- : bool -> 'a -> bool * (bool * 'a) = <fun>
+|}]
+
+let _ = fun a b -> match a, b with
+| (true as x, _) as p -> x, p
+| (false, x) as p -> x, p
+(* outside, complex *)
+[%%expect{|
+(function a/128 b/129
+  (if a/128
+    (let (x/130 =a a/128 p/131 =a (makeblock 0 a/128 b/129))
+      (makeblock 0 x/130 p/131))
+    (let (x/132 =a b/129 p/133 =a (makeblock 0 a/128 b/129))
+      (makeblock 0 x/132 p/133))))
+- : bool -> bool -> bool * (bool * bool) = <fun>
+|}]
+
+let _ = fun a b -> match a, b with
+| ((true as x, _) as p)
+| ((false, x) as p)
+  -> x, p
+(* inside, complex *)
+[%%expect{|
+(function a/134 b/135
+  (catch
+    (if a/134
+      (let (x/142 =a a/134 p/143 =a (makeblock 0 a/134 b/135))
+        (exit 10 x/142 p/143))
+      (let (x/140 =a b/135 p/141 =a (makeblock 0 a/134 b/135))
+        (exit 10 x/140 p/141)))
+   with (10 x/136 p/137) (makeblock 0 x/136 p/137)))
+- : bool -> bool -> bool * (bool * bool) = <fun>
+|}]
+
+(* here flattening is an optimisation,
+   as we avoid allocating the tuple in the first case,
+   and only allocate in the second case *)
+let _ = fun a b -> match a, b with
+| (true as x, _) as _p -> x, (true, true)
+| (false as x, _) as p -> x, p
+(* outside, onecase *)
+[%%expect {|
+(function a/144 b/145
+  (if a/144
+    (let (x/146 =a a/144 _p/147 =a (makeblock 0 a/144 b/145))
```

Comment:
Isn't the tuple being allocated here?

Reply:
It's subtle. The declaration shows up, but it is an unused alias, so it is removed by simplification: it is absent in the `-dlambda` output. I think that I wrote the original tests with `-dlambda`, and wrote the comments accordingly, then someone (my bet is on @trefis) asked for `-drawlambda` output, and I missed this small surprise. Thanks for noticing!

The testsuite is a bit hard to make sense of today, because it mostly shows things that are handled in the exact same way in all cases. Before the PR, it would show fairly different results; in particular, when an allocation happens due to Cannot_flatten, it is at the top of the matching (before any conditional test), and always used as the match scrutinee, so never erasable. Here we push the allocations within the branch, and in some cases they get erased by simplification (which is the point of trying harder to flatten).

The test that you are looking at is a test that was always flattenable. In particular, ocaml 4.10 and this PR produce the same `-dlambda` and the same `-drawlambda` output. The "outside, tuplist" test below is similar. But this is not the case for the "inside, tuplist" test below it, that now only allocates in one branch and not in the other (but the allocation shows in the `-drawlambda` output.)

It is not good that the test output is confusing in this way, but I am not sure how to fix it. We could:
- add a comment in the test
- revert to -dlambda reference output
- change the code generator to not generate a binding for a variable that is not used in the action anyway

I think I will start with just a comment, but I am not confident that this is the best approach. Any advice?

Reply:
Regarding the testsuite:  my point is that this test is not so useful to understand the behavior today, but rather to:
- show up as interesting diffs in the commit that implements the change of behavior in the PR
- test the space of various situations, and serve as regression tests if we evolve this part of the code in the future

It sounds believable to me that the same "differences between versions" that we had with Cannot_flatten could re-appear with another implementation; then we could decide whether we accept them or not, but at least we have the corner-case differences in the testsuite.

Reply:
The tuplist examples avoid the allocation on the first branch, aren't they? 

I would be happy with comments that match the result of the tests. If I am not mistaken, with the rawlambda output, we are testing that the tuple allocation happen inside the branch, when needed, and independently of the presence of an exploded-or.

Reply:
Before this PR (for example in 4.10), the "outside" test would avoid the allocation in the first branch, but the "inside" test would raise Cannot_flatten and allocate in both cases. After the PR, both tests avoid the allocation in the first branch.

Thanks for the suggestion on how to write the comment.

Reply:
I rephrased the comment, and also I decided to inculde *both* `-drawlambda` and `-dlambda` output, to make the situation less perplexing in the first place.

Reply:
This seems fine to me.